Updating the header file with DNS rules during policy update

vipul-21 opened the pull request:
Writing the DNS rules to the endpoint header/json file as we apply the policy changes, This keeps the ep_config.json and ep_config.h in sync with the current rules being applied the policy.

The PR is based on the comments of the PR: #33412 (Got closed due to inactivity so had to reopen it)

```release-note
Add flag to configure how frequently DNS information is stored to disk, with a default of waiting at least 5s and a maximum of 5m between disk write operations per Endpoint.
```

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2437,18 +2427,14 @@ func (e *Endpoint) syncEndpointHeaderFile(reasons []string) {
-	if trigger := e.dnsHistoryTrigger.Load(); trigger != nil {
-		trigger.Trigger()
-	}
+	e.controllers.TriggerController(e.GetK8sNamespaceAndPodName() + "-sync-endpoint-header-file")
```
This is the line I was missing on my initial glance. :+1:

joestringer replied:
(Will defer to other reviewers, I was just passing through)

asauber commented on pkg/controller/controller_test.go:
```diff
@@ -221,3 +221,26 @@ func TestWaitForTermination(t *testing.T) {
+	time.Sleep(time.Duration(1) * time.Second) // wait for controller to start running
+	firstTimeStamp := ctrl.lastSuccessStamp    // get first timestamp
+	mngr.TriggerController("test")             // trigger controller
+	time.Sleep(time.Duration(2) * time.Second) // wait for controller to run and update timestamp
```
Since you sleep here, you're not testing the value of `MinTriggerInterval`.

What this test should do is not sleep at all, expect the Controller to sleep, and then check that the total duration was more than 1900ms, not less than 2100ms.

vipul-21 replied:
So the `sleep` is allow the `lastSuccessStamp` to get update since `MinTriggerInterval` is 2 sec.

If I remove the sleep, the `lastSuccessTimestamp` does not get updated fast enough. For example,
```
 controller_test.go:238: First timestamp:  2025-01-10 23:18:27.522208261 +0000 UTC m=+0.006065412
 controller_test.go:242: Second timestamp:  2025-01-10 23:18:27.522208261 +0000 UTC m=+0.006065412
```

With the sleep
```
    controller_test.go:238: First timestamp:  2025-01-10 23:19:34.410793619 +0000 UTC m=+0.005924102
    controller_test.go:242: Second timestamp:  2025-01-10 23:19:36.411650727 +0000 UTC m=+2.006781110
```
Since the controller has a interval of 10 secs, triggering it again after the initial success should obey the `MinTriggerInterval` which is 2 secs in this case.

The total duration you are talking about is calculated after the initial delay(MinTriggerInterval). So that would not account for the minTriggerInterval.

```  
               if params.MinTriggerInterval > 0 {
			if delayNeeded, delay := c.needsDelay(params.MinTriggerInterval); delayNeeded {
				time.Sleep(delay)
			}
		}

		start := time.Now()
		err = params.DoFunc(params.Context)
		duration := time.Since(start)
```
I agree to add 
```require.Greater(t, timeDiff, time.Duration(1900)*time.Millisecond)``` along with the `Less` condition to make sure MinTriggerInterval is accepted.

asauber replied:
I will have to take another close look to respond properly here, but my quick thought is that as long as you have a `Sleep` in the context of the test, and then check if the delay was between 1.9 and 2.1 seconds, then you're not validating the minimum interval. Maybe based on what you are pointing out, you need to wait for an initial period, and then perform another trigger and observe that that second trigger also took at least 1.9 seconds with no "top level" sleep.

vipul-21 replied:
``` you have a Sleep in the context of the test, and then check if the delay was between 1.9 and 2.1 seconds, then you're not validating the minimum interval.```
I am not sure if this is the case. The sleep for 2 seconds is to imitate(make sure) that the controller after the second trigger gets executed correctly as the `MinTriggerInterval` is 2 seconds and updates the timestamp. Even if I add a sleep of 4 seconds(instead of 2 seconds), the time difference between the two runs should still be between 1.9 and 2.1 seconds as the 2nd trigger happened just after the first time stamp is updated. And as per the code it should wait for the `MinTriggerInterval`, if it does not then there is failure case. (Also the regular interval for controller is 10 seconds to make sure it does not get executed while we test MinTriggerInterval if that make sense)

If we don't have the sleep, we would need to add a counter for number of successful run. And then maybe compare the time diff between the runs with `MinTriggerInterval`. 

As per my initial comment, the trigger itself is not 1.9sec or more. It just adds the delay of `MinTriggerInterval` i.e. 2 secs. The logic for calculating the trigger duration is outside the delay logic I added.

asauber replied:
> So the sleep is allow the lastSuccessStamp to get update since MinTriggerInterval is 2 sec.

Got it, this sleep is to allow the test to observe the triggered interval. Not sure why I didn't read it this way earlier.

Yep, this looks good to me, if we also check `require.Greater(t, timeDiff, time.Duration(1900)*time.Millisecond)`

joestringer replied:
Can we test this without actually sleeping? When running tests in CPU constrained environments, relying on sleeps can become unreliable, then becoming a cause for flakiness.

vipul-21 replied:
Updated the test with the second option I mentioned above by using the success count.

asauber reviewed (changes requested):
Thanks for implementing the Interval + MinInterval concept. Since we know the Controller is on it's own goroutine, the implementation looks good to me.

I have one comment regarding your test.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2422,7 +2412,7 @@ func (e *Endpoint) IsDisconnecting() bool {
 }
 
-func (e *Endpoint) syncEndpointHeaderFile(reasons []string) {
+func (e *Endpoint) syncEndpointHeaderFile() {
```
I dug a little bit into why the reasons were dropped here - the goal was to allow logging different callsites, but in practice there's only one callsite. So the change makes sense even though it wasn't called out in the commit message :+1: 

Digging a couple of layers deeper, multiple callers may call into the underlying `writeHeaderfile` and trigger writes, but they are protected by a mutex. Strictly speaking this means it's possible to write the file more often than the controller dictates, but there's no bug where multiple writers concurrently write and break the file. The final step of that function is an atomic replace as well so there should not be any partial writes.

Maybe over time if we keep looking at this, it might be interesting to serialize the DNS information separately from the headerfile or better structure the headerfile write at least to better enforce a maximum rate of writes to disk. But I'm not overly concerned about this, and certainly not to the degree I'd block this PR. If you are curious to work on improving this, then feel free to send a follow-up PR to improve logic in this area.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1268,10 +1262,6 @@ func (e *Endpoint) leaveLocked(conf DeleteConfig) []error {
 	e.controllers.RemoveAll()
 	e.cleanPolicySignals()
 
-	if trigger := e.dnsHistoryTrigger.Swap(nil); trigger != nil {
-		trigger.Shutdown()
-	}
```
Where's the corresponding shutdown logic with the new implementation? Do we just rely on the controller call to run to completion?

joestringer replied:
Maybe the answer is that controller handles this for us - the controller would receive a context cancellation from the overall agent, then the question is exactly what happens to the trigger logic embedded in the controller and will that then become a no-op for a trigger initiated after shutdown begins? I didn't poke around far enough in the code to confirm.

vipul-21 replied:
For the shutdown logic. The controllers are removed during the deletion of the endpoint `e.controllers.RemoveAll()`, which will initiate a shutdown as you mentioned.

For the trigger that happens after the shutdown begins should follow the same pattern as before. The controller trigger was present already(just introduced a delay) and in case there is trigger after the shutdown we have this:
```
// If we receive a signal on multiple channels golang will pick one randomly.
// This select will make sure we don't execute the controller
// while we are shutting down.
select {
case <-c.stop:
    goto shutdown
default:
}
````

joestringer commented on pkg/controller/controller.go:
```diff
@@ -247,6 +252,12 @@ func (c *controller) runController(params ControllerParams) {
+			if delayNeeded, delay := c.needsDelay(params.MinTriggerInterval); delayNeeded {
+				time.Sleep(delay)
+			}
+		}
```
The child function already checks the zero condition, no need to do it here:
```suggestion
		if delayNeeded, delay := c.needsDelay(params.MinTriggerInterval); delayNeeded {
			time.Sleep(delay)
		}
```

joestringer commented on pkg/controller/controller.go:
```diff
@@ -352,6 +363,20 @@ shutdown:
+func (c *controller) needsDelay(minInterval time.Duration) (bool, time.Duration) {
+	if minInterval == time.Duration(0) {
+		return false, 0
+	}
```
nit: `time.Sleep(time.Duration(0))` returns immediately, so we can simplify by just skipping the bool.

```suggestion
func (c *controller) needsDelay(minInterval time.Duration) (time.Duration) {
	if minInterval == time.Duration(0) {
		return 0
	}
```

joestringer commented on pkg/controller/controller_test.go:
```diff
@@ -221,3 +221,40 @@ func TestWaitForTermination(t *testing.T) {
+			mngr.TriggerController("test")
+		}
+
+		if ctrl.GetSuccessCount() == 2 {
+			secondTimeStamp = ctrl.lastSuccessStamp
+			timeDiff := secondTimeStamp.Sub(firstTimeStamp)
+			require.Less(t, timeDiff, time.Duration(2100)*time.Millisecond)
+			require.Greater(t, timeDiff, time.Duration(1900)*time.Millisecond)
+		}
```
Is this condition ever hit by the test?

vipul-21 replied:
Yes it does.
```
   controller_test.go:245: First success timestamp:  2025-01-16 05:30:48.931834724 +0000 UTC m=+0.007504264
    controller_test.go:252: Second success timestamp:  2025-01-16 05:30:50.932870994 +0000 UTC m=+2.008540534
    ```

joestringer replied:
Oh I see, I misread the loop as having 3 iterations rather than 100 iterations in worst case (~= 10s). Hmm.

joestringer replied:
I see we're already doing effectively the same thing already in other tests in this file, so it's probably not reasonable to block the PR on coming up with a better solution to this specific piece. I also couldn't think of an obvious better way to do this off the top of my head. I don't love it, but I'm probably overthinking the risk. Marking as resolved.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -613,23 +611,20 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
-		return
-	}
-	e.dnsHistoryTrigger.Store(trigger)
+func (e *Endpoint) initEndpointSyncController(interval time.Duration) {
```
I would suggest documenting somewhere here that the point of this controller is to ensure timely sync of DNS information to disk. With the removal of the trigger named `dnsHistoryTrigger`, this context is lost from the codebase.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2442,6 +2442,10 @@ func (e *Endpoint) syncEndpointHeaderFile() {
 func (e *Endpoint) SyncEndpointHeaderFile() {
+	if e.controllers == nil {
+		return
+	}
```
Shouldn't this test properly initialize the object before using it? It bothers me to sprinkle error checks like this throughout active production code, because it implies that some part of the code is not properly initializing some structures. This can lead to higher cognitive overload on future developers/users because we haven't locked down the assumptions in the logic.

joestringer replied:
I suggest dropping this specific commit and replacing it with better initialization in the test code.

vipul-21 replied:
The check was added for the test case where the controller is not initialized. Otherwise, the controllers are initialized properly.
Updated the test case which was failing.(endpointsynchronizer_test.go)

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -616,8 +616,9 @@ func (e *Endpoint) initEndpointSyncController(interval time.Duration) {
-			Context:     e.aliveCtx,
+			RunInterval:        interval,
+			Context:            e.aliveCtx,
+			MinTriggerInterval: 5 * time.Second,
```
Could you squash this into the first commit?

vipul-21 replied:
Done

joestringer commented on pkg/controller/controller_test.go:
```diff
@@ -233,15 +233,28 @@ func TestTriggerController(t *testing.T) {
-
-	require.Less(t, timeDiff, time.Duration(2100)*time.Millisecond)
-	require.Greater(t, timeDiff, time.Duration(1900)*time.Millisecond)
+	var firstTimeStamp time.Time
```
Could you squash this into the second commit?

vipul-21 replied:
Done

joestringer commented on pkg/controller/controller.go:
Could you update the commit message for this last commit to explain why the change is being made? "Update .... function" does not explain your intent, so if someone later on finds a bug in the implementation they won't be able to reason about whether the bug is intentional behavior or a side-effect of the change being made here.

joestringer replied:
Oh, I think this is all in response to my review feedback. Could you instead fold the changes back into the original commits that are changing the behavior (probably the first commit)?

joestringer reviewed (commented):
I re-reviewed commit-by-commit, and was confused because it seemed like the feedback was not fixed. However, it looks like you have actually fixed the feedback, it's just that the changes weren't squashed back into the commits that the feedback was provided against. Would you mind fixing that up?

For context in Cilium we typically uphold a pretty high bar for git history cleanliness for a few reasons:
- Ensure git bisect provides a valid version of the logic to help with bisecting bugs in future
- Provide context to developers digging through the history
- Simplify review experience by collecting logically related changes all into a single commit so that the codebase transitions atomically from one version of the code to another.

As for logically related changes, I see two separate logical changes here: Extending the controller trigger logic, and changing the endpoint DNS rule sync from a trigger to a controller.

We recently discussed this [on Slack](https://cilium.slack.com/archives/C2B917YHE/p1736191830778769) as well and I think we might need to update the contribution guidelines to clarify this, but that will also maybe provide some additional context behind what I'm saying.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -609,23 +607,21 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
+				e.syncEndpointHeaderFile()
+				return nil
+			},
+			RunInterval:        interval,
```
Hmm, something else I realized while taking another look is that this will force the write to disk every 5m even if there's no new information to write. Right? This seems slightly unnecessary and not aligned with the intent to make the update every time policy changes are made. It doesn't seem obviously wrong to me per se but I'm just wondering if I'm missing something or if the switch to a controller is the right way to achieve the goal of this PR.

vipul-21 replied:
Yeah that is correct, this what the previous PR talked about as a solution: https://github.com/cilium/cilium/pull/33412#issuecomment-2225171133, I wonder if we should be able to trigger the old Trigger on the regeneration of endpoints ? Not super familiar with regenerate code but in that case we won't be updating every 5 minutes. But not sure if that is possible because as mentioned in the comment the dns pod ip can change irrespective of the endpoint life cycle change.
Or we can make the controller func smart enough to not trigger if there are no changes from the past rules ?

joestringer replied:
Ah I see, I was missing that context. Thanks for sharing that. I'm guessing that overall if we can track down all the points that mutate this state and get them a pointer to the endpoint, it should be possible to let them trigger the sync. On the other hand maybe if @squeed is recommending a periodic controller then that's a simpler / more reliable mechanism to make really sure it happens.

vipul-21 replied:
@joestringer Tried looking at the possibility of triggering the sync if the dns endpoint ip changes. I couldn't find a way where the dns endpoints ip change can reflect the update on the egress endpoint for which the policy is applied.(Atleast not any straight forward way)

Just a thought, we can even make the default as 0. In that case, the controller behaves as a trigger. (Old behavior) And for folks who need the updated endpoint dns rules, they can configure the value of `EndpointHeaderFileSyncInterval`

joestringer replied:
Hmm. Structurally we already embed a DNS history object in the endpoint, so it seems natural that we already have an association between the DNS entries and the endpoints that they impact, so then we should in theory be able to piggyback off that association to determine which endpoints need to sync their state to disk.

> And for folks who need the updated endpoint dns rules, they can configure the value of `EndpointHeaderFileSyncInterval`

Broadly, having a configuration option (even if hidden) can help power users to tweak Cilium to their needs, but I would rather not rely on the idea that people are changing the setting. Cilium's configuration space is complicated enough as it is. As developers I would hope we can come up with a single sane default that we believe is the right way for Cilium to operate in a wide range of deployments, and have that as the primary way we reason about Cilium's behavior. If we think that the code structure currently limits our ability to implement a better solution to this problem, then maybe the conclusion is to keep the periodic behavior by default. We can optimize it further later.

joestringer replied:
I'm not seeing a clear path to an obviously better option, so it probably makes sense to accept this change as-is and if necessary iterate to further improve. Blocking this PR further on a better solution to this particular problem doesn't seem like it makes sense at this point. Resolving.

vipul-21 committed:
Add min interval for the controller

Min Interval for the controller is the minimum interval between controller runs.
If the controller is triggered more frequently than this interval, the
controller will wait until the interval has passed before running.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

vipul-21 committed:
Updating header file during policy update using a controller

Previously the header file generation was triggered on every policy update,
but not more than once every five seconds. The controller is now triggered
once every five minutes and also on every policy update with same minimum interval

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>
